src/allocator.rs

```diff
@@ -152,6 +153,16 @@ impl Allocator {
         node_from_number(self, &v)
     }
 
+    pub fn new_g1(&mut self, g1: G1Projective) -> Result<NodePtr, EvalErr> {
+        let g1: G1Affine = g1.into();
```

Comment:
It's going to be a lot faster if you simply leave it as projective.

Reply:
great! I'll make that change

Reply:
oh, maybe I misunderstood you. Are you're referring to the serialization into bytes, or the API taking a `G1Projective` instead of `G2Projective`.

Would it make more sense for this API to take `G1Affine` or `G1Projective`? My assumption was that `G1Projective` is the most complete state, and most efficient to keep intermediate values in.

Reply:
Yes, conversion from `G1Projective` (X, Y, Z) to `G1Affine` (x, y) is costly. The other way is nearly free.

Operating on `G1Projective` is fastest, although uses more memory. You generally want to convert to `G1Affine` before serializing or comparing as `G1Affine` serializes smaller and has canonical representations.

You can have two `G1Projective` points that look very different but are the same. You convert to `G1Affine` to canonicalize.